File: 0112: pathSum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        
        if(root == nullptr){
        
            return false;
        }

        return valid(root, targetSum);
    }

    bool valid(TreeNode* root, int sum){

        if(root == nullptr){

            return false;
        }

        if(root->left == nullptr && root->right == nullptr){

            if(sum == root->val){

                return true;
            }

            return false;
        }
        sum -= root->val;

        return valid(root->left, sum) || valid(root->right, sum);

        
    }
};

File: 0112: pathSum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        
        def path(node, targ):
            if node is None:
                return False
            if node.left is None and node.right is None:
                if (node.val - targ) == 0:
                    return True
            return path(node.left, targ-node.val) or path(node.right, targ-node.val)
        return path(root, targetSum)